app/AvailablePlugin/PasswordAuthenticator/Model/PasswordResetToken.php

```diff
@@ -260,9 +251,60 @@ protected function send($passwordAuthenticator, $coPersonId, $token, $recipients
       'LINK_EXPIRY'       => $expiry
     );
 
-    // Construct subject and body
+    $this->send($passwordAuthenticator, $coPerson, $recipients, $actorCoPersonId, $substitutions);
+
+    // Also store the recipient list in the token
+    $this->clear();
+    $this->updateAll(
+      array('PasswordResetToken.recipients' => "'" . substr(implode(',', $recipients), 0, 256) . "'"),
+      array('PasswordResetToken.token' => "'" . $token . "'")
+    );
+  }
 
-    $msgSubject = processTemplate($mt['CoMessageTemplate']['message_subject'], $substitutions);
+  /**
+   * Send a Password Reset  or Username Reminder
+   *
+   * @since  COmanage Registry v4.1.0
+   * @param  PasswordAuthenticator $passwordAuthenticator Password Authenticator configuration
+   * @param  array                 $coPerson              CO Person with associated email addresses and identifiers
+   * @param  array                 $recipients            Array of email addresses to send token to
+   * @param  int                   $actorCoPersonId       Actor CO Person ID
+   * @param  array		   $substitutions	  substitutions for the message template
+   * @throws InvalidArgumentException
+   */
+  protected function send($passwordAuthenticator, $coPerson, $recipients, $actorCoPersonId, $substitutions=null) {
+
+    // If $substitutions are null, this is a username reminder. Set the message template type and text for history record, to be used later, appropriately based on this
+    if ($substitutions == null) {
```

Comment:
Checking for `$substitutions` being NULL is brittle. Instead, either pass a mode parameter, or have the caller provide the message template and history text.

Or since we already have two functions anyway (`send` and `sendRequest`), just have two separate functions - one for reset and the other for reminder.